presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPath.java

```diff
@@ -13,19 +13,146 @@
+            {
+                JsonNode node = jaywayExtract(jsonPath, jaywayConfig, inputStream);
+                if (node == null) {
+                    return null;
+                }
+                return utf8Slice(node.toString());
+            }
+        };
+    }
+
+    private static JsonExtract.JsonExtractor<Long> getSizeExtractorForJayway(com.jayway.jsonpath.JsonPath jsonPath, Configuration jaywayConfig)
+    {
+        return new JsonExtract.JsonExtractor<Long>()
+        {
+            @Override
+            public Long extract(InputStream inputStream)
+                    throws IOException
+            {
+                JsonNode node = jaywayExtract(jsonPath, jaywayConfig, inputStream);
+                if (node == null) {
+                    return null;
+                }
+                return (long) node.size(); // Jackson correctly returns 0 for scalar nodes
+            }
+        };
+    }
+
+    private static JsonNode jaywayExtract(com.jayway.jsonpath.JsonPath jsonPath, Configuration jaywayConfig, InputStream inputStream)
+            throws IOException
+    {
+        try {
+            Object res = jsonPath.read(inputStream, jaywayConfig);
+            if (res instanceof JsonNode) {
+                return (JsonNode) res;
+            }
+            else {
+                // Jayway will respect Jackson mappings as provided in the configuration and return a JsonNode for simple cases.
+                // But for JsonPath functions ($.avg, ...), it will return a Java boxed type (Double, String etc.) instead
+                // of a properly formed JsonNode. This is why we need to re-create a JsonNode in that case
+                return mapper.valueToTree(res);
+            }
+        }
+        catch (InvalidJsonException | PathNotFoundException ex) {
+            // replicate Presto's JsonPath behaviour: if the input JSON is invalid or no result is found,
+            // then return NULL instead of throwing an exception
+            return null;
+        }
+    }
+
+    public static JsonPath build(String pattern, JsonPathExtractionEngine engine)
+    {
+        switch (engine) {
+            case PRESTO:
+                return new JsonPath(JsonExtract.generateExtractor(pattern, new JsonExtract.ScalarValueJsonExtractor()),
+                        JsonExtract.generateExtractor(pattern, new JsonExtract.JsonValueJsonExtractor()),
+                        JsonExtract.generateExtractor(pattern, new JsonExtract.JsonSizeExtractor()));
+            case JAYWAY:
+                try {
+                    Configuration jaywayConfig = Configuration.builder().jsonProvider(new JacksonJsonNodeJsonProvider()).build();
+                    if (pattern == null || pattern.isEmpty()) {
+                        // for some reason, jayway throws IllegalArgumentException for an empty path, but an InvalidPathException for other invalid paths
+                        throw new InvalidPathException();
```

Comment:
What is the behavior is Presto json parser, when the pattern isnull or empty. Should it include an error message ?

Reply:
When the pattern is empty, presto throws `PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid JSON path: '')`. This is replicated by catching InvalidPathException and re-throwing a PrestoException on line 133 below.